AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m

```diff
@@ -67,23 +67,29 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupId:(NSString *)groupId flags:(MSFlags)fl
                                    targetKey ? [NSString stringWithFormat:@"'%@'", targetKey] : @"NULL", (unsigned int)persistenceFlags];
   }
   return [self executeQueryUsingBlock:^int(void *db) {
+           // Selecting logs with equal or lower priority and ordering by priority then age.
            int result = [MSDBStorage executeNonSelectionQuery:addLogQuery inOpenedDatabase:db];
 
            // If the database is full, delete logs until there is room to add the log.
            long countOfLogsDeleted = 0;
-           while (result == SQLITE_FULL) {
-             result = [MSLogDBStorage deleteOldestLogsWithCount:1 inOpenedDatabase:db];
+           NSArray<NSNumber *> *ids = [self logsWithPriorityLowerThan:persistenceFlags inOpenedDatabase:db];
```

Comment:
In the default case when this method is being invoked, we will be loading around 10 mb of data (since that is the default max storage size). In the typical case, we will be loading even more (since we expect devs to increase it for their app). Further, we will only really need to delete 1 or 2 logs on average to make room. 

Thus, I think we are loading several megabytes of data in memory when we probably need only a couple of kilobytes at the most. This seems very inefficient.

I propose that we load the logs in chunks of, say, 5 (but that's arbitrary). The previous implementation loaded 1 at a time, which would still be much more efficient. Also, the Android implementation seems to load logs individually.

Does anyone have thoughts on this?

Reply:
Yeah I think that this would be a good optimization.  If we're not concerned with matching Android exactly, I think that loading a few 5-10 logs at a time would be good.  

Is it also worth testing the worst-case scenarios to see if/how much the performance decreases if we have to load all the logs anyways?  I would be a bit concerned about the case where we have a DB full of critical logs and we try to add a normal one.  I think there are probably some customers who will run into this, and if it's much worse we should use larger chunks.

Reply:
I would say we get chunks that are already filtered by priority, so in the latter case the query would return 0 instead of 5

Reply:
Bsically just adding a count parameter to the method would do the trick

Reply:
This method doesn't load whole logs. It only returns database ID which is a primary keys for logs. It should be okay.

Reply:
I do share the concern though, can we do this to match Android cursor? https://stackoverflow.com/a/12387605/598520 the step thing.

Reply:
@jaelim-ms that's true, my mistake. but it could still be a lot so I think it might be worth changing.

Reply:
Even 1 at a time should do the trick but I would check the >= priority only if the database is full. Having the DB full is a bit of an edge for when the connection in down or advanced scenario where the transmission is paused for a while and even then the DB needs to reach its full capacity.

Reply:
@clpolet That was considered while implementing it but I forgot to address. 😄 
@achocron It is even worse without this call. Deleting one log at a time will do the full scan anyway. I'll change this to call only when the db is full.